Update collection-expressions.md (#7481)
dotnet/csharplang: proposals/csharp-12.0/collection-expressions.md

### Spec clarifications
[spec-clarifications]: #spec-clarifications

> * For brevity, `collection_expression` will be referred to as "literal" in the following sections.

Reviewer: Probably would reword this to remove references to `literal`

## Conversions
[conversions]: #conversions

A *collection expression conversion* allows a collection literal expression to be converted to a type.

The following implicit *collection expression conversions* exist from a collection literal expression:

> * To a single dimensional *array type* `T[]` where:

Author: ```suggestion
* To a single dimensional *array type* `T[]` where:

```

* To a single dimensional *array type* `T[]` where:

> 

Reviewer 2: The blank lines were added in https://github.com/dotnet/csharplang/pull/7468, and might be necessary.

Author: ok :)

> * To a *span type* `System.Span<T>` or `System.ReadOnlySpan<T>` where:

Author: ```suggestion
* To a *span type* `System.Span<T>` or `System.ReadOnlySpan<T>` where:

```

  * For each *element* `Ei` there is an *implicit conversion* to `T`.

> * To a *type* with a *[create method](#create-methods)* with *parameter type* `System.ReadOnlySpan<T>` where:

Author: ```suggestion
* To a *type* with a *[create method](#create-methods)* with *parameter type* `System.ReadOnlySpan<T>` where:

```

  * For each *element* `Ei` there is an *implicit conversion* to `T`.

> * To an *[inline array type](https://github.com/dotnet/csharplang/blob/main/proposals/inline-arrays.md)* with *element type* `T` where:

Author: ```suggestion
* To an *[inline array type](https://github.com/dotnet/csharplang/blob/main/proposals/inline-arrays.md)* with *element type* `T` where:

```

## Conversions
[conversions]: #conversions

> A *collection expression conversion* allows a collection literal expression to be converted to a type.

Reviewer: ```suggestion
A *collection expression conversion* allows a collection expression to be converted to a type.
```

> The following implicit *collection expression conversions* exist from a collection literal expression:

Reviewer: ```suggestion
The following implicit *collection expression conversions* exist from a collection expression:
```

>     Here, if `b` is false, it is not required that any value actually be constructed for the empty literal since it would immediately be spread into zero values in the final literal.

Reviewer: ```suggestion
    Here, if `b` is false, it is not required that any value actually be constructed for the empty collection expression since it would immediately be spread into zero values in the final literal.
```

> * The empty literal is permitted to be a singleton if used to construct a final collection value that is known to not be mutable.  For example:

Reviewer: ```suggestion
* The empty collection expression is permitted to be a singleton if used to construct a final collection value that is known to not be mutable.  For example:
```

  * If `T` is some `ReadOnlySpan<T1>`, then the literal is translated the same as for the `Span<T1>` case except that the final result will be that `Span<T1>` [implicitly converted](https://learn.microsoft.com/dotnet/api/system.span-1.op_implicit#system-span-1-op-implicit(system-span((-0)))-system-readonlyspan((-0))) to a `ReadOnlySpan<T1>`.

    A `ReadOnlySpan<T1>` where `T1` is some primitive type, and all collection elements are constant does not need its data to be on the heap, or on the stack.  For example, an implementation could construct this span directly as a reference to portion of the data segment of the program.

>     The above forms (for arrays and spans) are the base representations of the literal value and are used for the following translation rules:

Reviewer 2: ```suggestion
    The above forms (for arrays and spans) are the base representations of the collection expression and are used for the following translation rules:
```

  </details>

* Can/should the compiler emit `Array.Empty<T>()` for `[]`?  Should we mandate that it does this, to avoid allocations whenever possible?

    Yes. The compiler should emit `Array.Empty<T>()` for any case where this is legal and the final result is non-mutable.  For example, targeting `T[]`, `IEnumerable<T>`, `IReadOnlyCollection<T>` or `IReadOnlyList<T>`.  It should not use `Array.Empty<T>` when the target is mutable (`ICollection<T>` or `IList<T>`).

* Should we expand on collection initializers to look for the very common `AddRange` method? It could be used by the underlying constructed type to perform adding of spread elements potentially more efficiently.  We might also want to look for things like `.CopyTo` as well.  There may be drawbacks here as those methods might end up causing excess allocations/dispatches versus directly enumerating in the translated code.

>     Yes.  An implementation is allowed to utilize other methods to initialize a collection value, under the presumption that these methods have well-defined semantics, and that collection types should be "well behaved".  In practice though, an implementation should be cautious as benefits in one way (bulk copying) may come with negative consequences as well (for example, boxing a struct collection).

Reviewer 2: I think the question of which methods the compiler can and should use is still an open question. And we need to specific with the set of methods that might be used.

Author: Can we have the spec say that this is allowed, but have it be a compiler investigation as to what we actually use?

>     An implementation should take advantage in the cases where there are no downsides (like no boxing).  For example, with a `.AddRange(ReadOnlySpan<T>)` method.

Author: ```suggestion
    An implementation should take advantage in the cases where there are no downsides.  For example, with a `.AddRange(ReadOnlySpan<T>)` method.
```

    Yes. The compiler should emit `Array.Empty<T>()` for any case where this is legal and the final result is non-mutable.  For example, targeting `T[]`, `IEnumerable<T>`, `IReadOnlyCollection<T>` or `IReadOnlyList<T>`.  It should not use `Array.Empty<T>` when the target is mutable (`ICollection<T>` or `IList<T>`).

* Should we expand on collection initializers to look for the very common `AddRange` method? It could be used by the underlying constructed type to perform adding of spread elements potentially more efficiently.  We might also want to look for things like `.CopyTo` as well.  There may be drawbacks here as those methods might end up causing excess allocations/dispatches versus directly enumerating in the translated code.

    Yes.  An implementation is allowed to utilize other methods to initialize a collection value, under the presumption that these methods have well-defined semantics, and that collection types should be "well behaved".  In practice though, an implementation should be cautious as benefits in one way (bulk copying) may come with negative consequences as well (for example, boxing a struct collection).

>     An implementation should take advantage in the cases where there are no downsides.  For example, with a `.AddRange(ReadOnlySpan<T>)` method.

Author: ```suggestion
    An implementation should take advantage in the cases where there are no downsides.  For example, with an `.AddRange(ReadOnlySpan<T>)` method.
```
